File: day12/main.ts
import * as fs from 'fs';

const lines = fs.readFileSync('day12/input.txt').toString().split('\n');
console.error(`removing empty last line: ${lines.pop()}`);

interface Cmd {
  cmd: 'N'|'S'|'E'|'W'|'L'|'R'|'F';
  by: number;
}

function route(cmds: Cmd[]) {
  const pos = {x: 0, y: 0};
  const dirs = [{x: 1, y: 0}, {x: 0, y: -1}, {x: -1, y: 0}, {x: 0, y: 1}];
  let curDir = 0;
  for (const {cmd, by} of cmds) {
    switch (cmd) {
      case 'N': {
        pos.y += by;
        break;
      }
      case 'S': {
        pos.y -= by;
        break;
      }
      case 'W': {
        pos.x -= by;
        break;
      }
      case 'E': {
        pos.x += by;
        break;
      }
      case 'L': {
        const minus = by / 90;
        curDir = (((curDir - minus) % 4) + 4) % 4;
        break;
      }
      case 'R': {
        const plus = by / 90;
        curDir = (curDir + plus) % 4;
        break;
      }
      case 'F': {
        pos.x += dirs[curDir].x * by;
        pos.y += dirs[curDir].y * by;
        break;
      }
    }
  }
  return pos;
}

const dirs: Cmd[] = lines.map(line => {
  return {
    cmd: line[0],
    by: Number(line.substring(1)),
  } as Cmd;
});

console.error(`Part 1:`);
{
  const {x, y} = route(dirs);
  console.log(Math.abs(x) + Math.abs(y));
}

function route2(cmds: Cmd[]) {
  const pos = {x: 0, y: 0};
  const waypoint = {x: 10, y: 1};
  for (const {cmd, by} of cmds) {
    switch (cmd) {
      case 'N': {
        waypoint.y += by;
        break;
      }
      case 'S': {
        waypoint.y -= by;
        break;
      }
      case 'E': {
        waypoint.x += by;
        break;
      }
      case 'W': {
        waypoint.x -= by;
        break;
      }
      case 'L': {
        for (let i = 0; i < by / 90; ++i) {
          const ny = waypoint.x;
          const nx = -waypoint.y;
          waypoint.x = nx;
          waypoint.y = ny;
        }
        break;
      }
      case 'R': {
        for (let i = 0; i < by / 90; ++i) {
          const ny = -waypoint.x;
          const nx = waypoint.y;
          waypoint.x = nx;
          waypoint.y = ny;
        }
        break;
      }
      case 'F': {
        pos.x += waypoint.x * by;
        pos.y += waypoint.y * by;
        break;
      }
    }
  }
  return pos;
}

console.error(`Part 2:`);
{
  const {x, y} = route2(dirs);
  console.log(Math.abs(x) + Math.abs(y));
}

File: day12/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

type cmd struct {
	nm byte
	by int
}

type position struct {
	x int
	y int
}

func rleft(p position, amt int) position {
	for i := 0; i < amt; i += 1 {
		nx := -p.y
		ny := p.x
		p.x = nx
		p.y = ny
	}
	return p
}

func rright(p position, amt int) position {
	for i := 0; i < amt; i += 1 {
		nx := p.y
		ny := -p.x
		p.x = nx
		p.y = ny
	}
	return p
}

func route1(cmds []cmd) position {
	pos := position{x: 0, y: 0}
	dir := position{x: 1, y: 0}
	for _, cmd := range cmds {
		switch cmd.nm {
		case 'N':
			pos.y += cmd.by
		case 'S':
			pos.y -= cmd.by
		case 'E':
			pos.x += cmd.by
		case 'W':
			pos.x -= cmd.by
		case 'L':
			dir = rleft(dir, cmd.by/90)
		case 'R':
			dir = rright(dir, cmd.by/90)
		case 'F':
			pos.x += dir.x * cmd.by
			pos.y += dir.y * cmd.by
		}
	}
	return pos
}

func route2(cmds []cmd) position {
	pos := position{x: 0, y: 0}
	wp := position{x: 10, y: 1}
	for _, cmd := range cmds {
		switch cmd.nm {
		case 'N':
			wp.y += cmd.by
		case 'S':
			wp.y -= cmd.by
		case 'E':
			wp.x += cmd.by
		case 'W':
			wp.x -= cmd.by
		case 'L':
			wp = rleft(wp, cmd.by/90)
		case 'R':
			wp = rright(wp, cmd.by/90)
		case 'F':
			pos.x += wp.x * cmd.by
			pos.y += wp.y * cmd.by
		}
	}
	return pos
}

func abs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}

func distMan(p position) int {
	return abs(p.x) + abs(p.y)
}

func main() {
	fi, err := ioutil.ReadFile("day12/input.txt")
	check(err)
	ficontent := string(fi)
	lines := strings.Split(ficontent, "\n")
	lines = lines[:len(lines)-1]

	cmds := []cmd{}
	for _, line := range lines {
		by, err := strconv.Atoi(line[1:])
		check(err)
		nm := line[0]
		cmds = append(cmds, cmd{nm, by})
	}

	fmt.Print("Part 1:\n")
	fmt.Println(distMan(route1(cmds)))

	fmt.Print("Part 2:\n")
	fmt.Println(distMan(route2(cmds)))
}
